source/adaptive_load/metrics_evaluator_impl.cc

```diff
@@ -0,0 +1,122 @@
+  }
+  return evaluation;
+}
+
+const std::pair<const std::vector<const MetricSpec*>,
+                const absl::flat_hash_map<const MetricSpec*, const ThresholdSpec*>>
+MetricsEvaluatorImpl::ExtractMetricSpecs(
+    const nighthawk::adaptive_load::AdaptiveLoadSessionSpec& spec) const {
+  std::vector<const MetricSpec*> metric_specs;
+  absl::flat_hash_map<const MetricSpec*, const ThresholdSpec*> threshold_spec_from_metric_spec;
+  for (const MetricSpecWithThreshold& metric_threshold : spec.metric_thresholds()) {
+    metric_specs.push_back(&metric_threshold.metric_spec());
+    threshold_spec_from_metric_spec[&metric_threshold.metric_spec()] =
+        &metric_threshold.threshold_spec();
+  }
+  for (const MetricSpec& metric_spec : spec.informational_metric_specs()) {
+    metric_specs.push_back(&metric_spec);
+    threshold_spec_from_metric_spec[&metric_spec] = nullptr;
+  }
+  return {metric_specs, threshold_spec_from_metric_spec};
+}
+
+absl::StatusOr<nighthawk::adaptive_load::BenchmarkResult>
+MetricsEvaluatorImpl::AnalyzeNighthawkBenchmark(
+    const nighthawk::client::ExecutionResponse& nighthawk_response,
+    const nighthawk::adaptive_load::AdaptiveLoadSessionSpec& spec,
+    const absl::flat_hash_map<std::string, MetricsPluginPtr>& name_to_custom_metrics_plugin_map)
+    const {
+  if (nighthawk_response.error_detail().code() != static_cast<int>(absl::StatusCode::kOk)) {
+    return absl::Status(static_cast<absl::StatusCode>(nighthawk_response.error_detail().code()),
+                        nighthawk_response.error_detail().message());
+  }
+
+  nighthawk::adaptive_load::BenchmarkResult benchmark_result;
+  *benchmark_result.mutable_nighthawk_service_output() = nighthawk_response.output();
+
+  // A map containing all available MetricsPlugins: preloaded custom plugins shared across all
+  // benchmarks, and a freshly instantiated builtin plugin for this benchmark only.
+  absl::flat_hash_map<std::string, MetricsPlugin*> name_to_plugin_map;
+  for (const auto& name_plugin_pair : name_to_custom_metrics_plugin_map) {
+    name_to_plugin_map[name_plugin_pair.first] = name_plugin_pair.second.get();
+  }
+  auto builtin_plugin =
+      std::make_unique<NighthawkStatsEmulatedMetricsPlugin>(nighthawk_response.output());
+  name_to_plugin_map["nighthawk.builtin"] = builtin_plugin.get();
+
+  std::pair<std::vector<const MetricSpec*>,
+            absl::flat_hash_map<const MetricSpec*, const ThresholdSpec*>>
+      metric_spec_list_map = ExtractMetricSpecs(spec);
+  // MetricSpecs in original order of definition.
+  std::vector<const MetricSpec*> metric_specs = metric_spec_list_map.first;
+  // Pointer to the corresponding ThresholdSpec, or nullptr for informational metrics.
+  absl::flat_hash_map<const MetricSpec*, const ThresholdSpec*> threshold_spec_from_metric_spec =
+      metric_spec_list_map.second;
+
+  std::vector<std::string> errors;
+  for (const MetricSpec* metric_spec : metric_specs) {
```

Comment:
Reading this, I'm still wondering; it seems to me that having `ExtractMetricSpecs` return a `vector<std::pair<MetricSpec*, ThresholdSpec*>>` could simplify that interface as well as the code here; wouldn't that avoid the map lookups? The map usage would be in implementation detail in ExtractMetricSpecs, and not propagate outside of it. Of course, that might thwart re-use for other purposes perhaps.

Reply:
Oh that's a good idea, it preserves the order and is simpler. I definitely don't need the map. If someone did need a map later, I think they could pass the vector of pairs into a map constructor.

Reply:
Updated now to return a vector of pairs.